design/metadata-handling.md

```diff
@@ -0,0 +1,456 @@
+      * [Proposal](#proposal)
+         * [User Stories [optional]](#user-stories-optional)
+            * [Story 1](#story-1)
+            * [Story 2](#story-2)
+         * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-15T11:41-05:00 -->
+
+<!--te-->
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+Cloud-init templates offer a very powerful feature to render the configuration
+at runtime based on metadata given by Ironic. We want to take advantage of this
+feature to render node specific configuration for nodes in a machine deployment
+or in a control plane kubeadm object. In addition, we want to be able to
+generate node specific network configuration in CAPM3 based on a template.
+
+This design proposal introduces a new data template object that contains
+templates for network data and metadata secret generation per node. Once
+generated, the secrets are passed to Ironic for each node of a machine
+deployment or control plane.
+
+This proposal introduces a new `metaData` field in BareMetalHost and in
+Metal3Machine, a new `networkData` field in
+Metal3Machine and a new `dataTemplate` field in Metal3machine that allows
+generating the metadata and network data secrets content.
+
+## Motivation
+
+When deploying a machine deployment or a control plane with CAPI (v1alpha3), the
+KubeadmConfig applied is the same for all the nodes. However, we would want
+control on some fields, such as the node name for example. This can be a
+variable in the cloud-init template, but it needs to be provided through Ironic
+as metadata. And the metadata needs to be different for each of the nodes. We
+might also want some varying network configuration per node.
+For example we would want the nodes of a machine deployment to have names like
+`worker-np1-0`, `worker-np1-1`, `worker-np1-2` etc. and those names to be
+unrelated with the BareMetalHost names (that might be `machine-xyz`, reflecting
+some hardware related information, not Kubernetes node information).
+Hence we want a way to generate a node-specific metadata and network data for a
+machine deployment or Kubeadm Control Plane and pass it down to Ironic.
+
+### Goals
+
+The goals are :
+
+- to add a `metaData` field in BareMetalHost to have it configurable
+  for Ironic, and in the Metal3Machine for CAPM3 to be able to pass it down to
+  BareMetalHost.
```

Comment:
We should describe how we're going to merge the values that come from the template with the set of values the BMO provides on its own. The template can override values, but if there is no template or the template doesn't include one of the 4-5 things we always send then those values will still be there.

Reply:
okay, I will add the description. We will always set by default the UUID, we could maybe get rid of the rest of the default values, as they were a temporary workaround for until we have this feature